src/psyclone/gocean1p0.py

```diff
@@ -1865,3 +1877,118 @@ def data_on_device(self, parent):
+        # meta-arguments could be used to check them and also improve the
+        # generated code. (Issue #288)
+
+        # Error checking
+        if len(self.symbol_table.argument_list) < 2:
+            raise GenerationError(
+                "GOcean Kernel should always have at least two "
+                "arguments representing the iteration indices.")
+
+        if (self.symbol_table.argument_list[0].datatype != "integer"
+                or len(self.symbol_table.argument_list[0].shape) != 0):
+            raise GenerationError(
+                "GOcean kernel first argument should be a scalar integer.")
+
+        if (self.symbol_table.argument_list[1].datatype != "integer"
+                or len(self.symbol_table.argument_list[1].shape) != 0):
+            raise GenerationError(
+                "GOcean kernel second argument should be a scalar integer.")
+
+        # Declare OpenCL kernel
+        code = self.indent(indent)
+        code = code + "__kernel "
+        code = code + "void " + self._name + "(\n"
+
+        # Generate kernel arguments, first 2 iteration indices are implicit.
+        array_arguments = []
+        for symbol in self.symbol_table.argument_list[2:]:
+            code = code + self.indent(indent + 1)
+            if len(symbol.shape) > 0:
+                code = code + "__global "
+
+            if symbol.datatype == "real":
+                code = code + "double "
+            elif symbol.datatype == "integer":
+                code = code + "int "
+            elif symbol.datatype == "character":
+                code = code + "char "
+            else:
+                raise NotImplementedError(
+                    "Type '{0}' OpenCL declaration not supported."
+                    "".format(symbol.datatype))
+
+            if len(symbol.shape) > 0:
+                code = code + "* restrict "
+                array_arguments.append((symbol.name, len(symbol.shape)))
+
+            code = code + symbol.name + ",\n"
+
+        code = code[:-2]   # Remove last ",\n"
+        code = code + "\n" + self.indent(indent + 1) + "){\n"
+
+        # Declare local variables.
+        for symbol in self.symbol_table._symbols.values():
+            if symbol not in self.symbol_table.argument_list:
+                code = code + self.indent(indent + 1)
+                if symbol.datatype == 'real':
+                    code = code + "double "
+                elif symbol.datatype == 'integer':
+                    code = code + "int "
+                elif symbol.datatype == 'character':
+                    code = code + "char "
+                else:
+                    raise NotImplementedError(
+                        "Type '{0}' OpenCL declaration not supported."
+                        "".format(symbol.datatype))
+
+                if len(symbol.shape) > 0:
+                    code = code + "* "
+                code = code + symbol.name + ";\n"
+
+        # Generate a LEN variable for each array.
```

Comment:
I didn't know about NameSpaceManager, I added a condition to check for clashes in the SymbolTable.
I believe the NameSpace should be managed from the SymbolTable itself, I will add an issue to explore this.

Reply:
Thanks for creating the new issue. I'm not sure whether it makes sense to have the SymbolTable manage the name space or for it just to be the record of the existing variables. I guess it could have a method that wrapped the NameSpaceManager, e.g. the `declare` method could be responsible for ensuring that the variable name is unique.